### PR TITLE
The A.G.R Complex (Gateway Mission)

### PR DESCRIPTION
**Pull request by AyyRobotics**
# About The Pull Request

This PR adds the code, sprites, and map for a new gateway mission, known in the files as the AGRComplex.

**// Main Info**
The AGRComplex mission involves an explorer starting in the far left corner of the map in an outpost held by friendly NT Simplemobs, and must progress linearly throughout the facility, fighting through checkpoints and defector-held facilities while attempting to reach the opposite corner of the map, which contains the major reward. There is one main route to take, however there are multiple points where the explorer can choose an alternate path with different threats but a reward, either with an easier time travelling or help in an upcoming fight.

**// Lore**
While light, there are some papers and holodisks present throughout the map, telling the story of what happened on the complex and what occurred in the aftermath. I did this since it felt like few missions had a story attached to it.

**// NPCs**
The mission introduces 8 new human simplemobs, including 2 NT defectors, a Russian mercenary and commander, 2 cult soldiers, and 2 mostly-friendly NT personnel. Most of these are simple enemies made from the base Russian enemies, although with unique factions which allows them to fight eachother.

**// Map**
The map itself takes place in the Archeological and Geological Research Complex, or A.G.R Complex for short. A facility in a mountain range, with a cold, mildly damaging atmosphere that is otherwise safe to breathe. The map is mostly dark, but nearly the whole map is dimly lit, via the snow in the code, with a faint blue light to make seeing possible but difficult. There is also a safe spot near the entrance that gives the roleplay-heavy players a location to visit with minimal risk.

**// Rewards**
There are a wide variety of rewards spanning the entire map, from medical items and basic weapons/ammo to sustain an explorer while traveling, to research items like tech disks and BEPIS research to supplement them, to unique items like genetics chromosomes, a prototype cloner machine board, and even a mining hardsuit. The reason there is so many varied rewards is because I feel as though Gateway Missions are generally not worth the effort, often taking a lot of resources and time for minimal payoff, especially to the rest of the station who often get nothing.

**// Variants**
The AGRComplex is actually two maps, as there are two alternate versions of the map, each with it's own unique enemy, final reward, and cult theme. This is to hopefully have an equal level of appreciation for those who enjoy each cult specifically, and helps the map have more replay value.

**// Ghostrole**
The map also includes a dedicated ghostrole, known as the Cultist Survivor. They spawn at the opposite corner of the map from the gateway and are essentially tasked with the same as the explorers, except in reverse. As the name implies, they are cultists and are neutered versions of their respective cult, and directed to not be primary antagonists. There are two variants of these ghostroles, with Clockwork Cultists for the Clockwork map variant, and Blood Cultists for the Narsian map variant. I did this for two reasons: 1. Antagonists like Cultists are very rarely used in-game, and this allows them to be used in a way that doesn't impact the main round. And 2. Antagonists don't generally get an area to practice and learn their abilities, so this could serve as a way for people to learn how they work in a safe-ish environment.

**// Primary Reward _(Spoilers)_**
The main reward, achieved by the explorers reaching the opposite corner of the map is a plant. A map-specific plant known as either Matrigrass or Sanguintare. These plants are fully functional for botany, and can destructively analyzed to give a map-specific tech, or ground into a new reagent, either Empowered Zinc or Exalted Hemogen, which is used to create their items.

**// Clockwork Engineering**
Clockwork Engineering is the tech given by analyzing Matrigrass, and is used to print a variety of Clockwork-cult themed items. This includes:

_1. Brass and Brass Ingots (Smithing): Same as their original variant.
2. Sentinel's Stimulant: A one-use, portable item that replicates the Sentinel's Compromise spell.
3. Tinderbox Trident: A ranged weapon (laser gun base) that fires the Clockwork Cult's 'Kindle' projectile with longer range.
4. G.E.A.R Hardsuit: A relatively protective and expensive hardsuit with a bright lamp.
5. Clockwork Parts: A new tier of Stock Parts, barring a power cell.
6. Clockwork Armament Module: A cyborg upgrade that turns cyborgs into neutered Clockwork Cultists, fit with their specific items and a special lawset to prevent griefing. When the upgrade is removed, the effect disappears._

**// Occult Designs**
Occult Designs is the tech given by analyzing Sanguintare. The items it gives include:

_1. Runed Metal, Runed Ingot, Construct Shells, and Imitation Soulstone Shards: Same as their original variants. Now, purified constructs can actually exist!
2. Eldritch Constructor: Just an upgraded RCD with improved effeciency.
3. Flagellants Cuirass: A variant of the Flagellant's Robe, trading some of the speed for minor armor, about the same as Durathread.
4. Crystalline Wire: A beartrap variant that also wounds the target it steps on, but is far more visible._

**// Food and Recipes**
Another thing that can be found by completing the mission is a map-specific recipe book, and 4 new food items using their respective cult reagents. It's strange, yes, but it gives the chef some variety.

## Why It's Good For The Game

The AGRComplex and it's additions not only act as a new gateway mission, giving gateway explorers a new area with interesting threats and rewards, but also giving ghostrole players a new and interesting option, and in some way aiding every single department.

Research gets boosts of money, illegal tech, BEPIS research, and the T6 parts to enjoy.

Medical gets excess medical supplies, genetics chromosomes, the sentinel stimulant, and a prototype cloner which is likely to result in some enjoyable RP opportunities.

Engineering gets excess materials and new equipment to use in the form of the GEARsuit and Cult RCD.

Security get a variety of new weapons and clothing, like the Tinderbox Trident and Flagellants Cuirass.

Service gets new plants for the Botanist to make and new meals for the Chef to cook.

Cargo gets money and possibly new tools to make use of, with help from Engineering. It also gives the miners something to do after mining!

Even Command gets some interesting RP opportunities as they may have to negotiate with the Cultist Survivors or report to Central what happened at the facility.

Overall, at least in one way, the mission seeks to benefit each member of the station. Explorers bring back things that help every department instead of just hoarding weapons for themselves, ghostrole enjoyers get an interesting new role to try while also letting them learn how Cultists play, and RP-focused players get new items to use in nearly every department, while having new opportunities for how to use them.

## A Port?

Nope.

## Pre-Merge Checklist

- [x] You tested this on a local server.
- [x] This code did not runtime during testing.
- [x] You documented all of your changes.

## Changelog

:cl:
add: Added the A.G.R Complex, two gateway missions with cult-themed gameplay and rewards.
add: Added the Cultist Survivor ghostrole, taking place on the AGRComplex gateway mission.
config: Adds AGRComplex file to config.
/:cl:

### CODE REVIEW

**Comment by AyyRobotics** [comment]
Note, there was a runtime error that occurred when attempting to start the gateway, something to do with how it rendered the inside image, but that also occurred when attempting to load other missions, so it's likely not caused by my code.

**Comment by AyyRobotics** [comment]
Same issue as the Engineering Meka PR had, now it's the error is in create_and_destroy with it hard deleting some shuttle parts.

**Comment by BongaTheProto** [comment]
Would it be possible to modularize the non-map code?

**Comment by AyyRobotics** [comment]
> Would it be possible to modularize the non-map code?

For the atmospherics.dm change? I could try doing that! If you also mean the sprites, I could do that too, though it would take some time to get everything together.

**Comment by AyyRobotics** [comment]
Moved COLD_ATMOS to the mission file, and it worked just fine!

**Comment by BongaTheProto** [comment]
Do you have a discord?

**Comment by BongaTheProto** [comment]
Do you have a discord?

**Comment by AyyRobotics** [comment]
> Do you have a discord?

I do, though I'd prefer not to use it if there's only a few changes needed!

**Comment by BongaTheProto** [comment]
> > Would it be possible to modularize the non-map code?
> 
> For the atmospherics.dm change? I could try doing that! If you also mean the sprites, I could do that too, though it would take some time to get everything together.

I mean it as in moving the new created icons and new .dm files to modular_splurt, while still allowing the code to work. Just wondering if that's possible

**Comment by AyyRobotics** [comment]
> I mean it as in moving the new created icons and new .dm files to modular_splurt, while still allowing the code to work. Just wondering if that's possible

I probably couldn't change anything in the config or .tme file, but the mission code and maps would work fine! Icons should also work, but it may take longer since I'll need to add some new sprite files, since modular_splurt doesn't have a few that the base has. If it'd work better that way, I can likely get it done by tomorrow!

**Comment by AyyRobotics** [comment]
> This seems AMAZING, would be lovely modularized for better code maintenance <3

Already got most of it modularized now! Just testing it locally to make sure I didn't miss anything. I can probably have the changes committed in an hour or so!

**Comment by AyyRobotics** [comment]
That should be all of it? I did it manually so it was a bit slow, but I shouldn't have missed or misplaced anything.

**Comment by AyyRobotics** [comment]
Realized that there were some issues loading the map manually, and also realized that fixing it was as simple as removing a hash in the awaymissionconfig file. Next commit will remove the hashes from the Complex maps for testing purposes!

**Comment by AyyRobotics** [comment]
Realized the new commit to fix testing doesn't actually work for some unknown reason. Going to try and fix it later today

**Comment by AyyRobotics** [comment]
Looked into fixing it, and in all honesty I have no idea what's even wrong in the first place. The right file, awaymissionconfig was modified, and that should be the file that controls which map gets selected. For some reason the only map that should be selected is jungleresort, but clearly in-game that's not the case, so either the game isn't actually registering the new changes to the config, or is just running off of an entirely different config file that I haven't found yet.

**Comment by AyyRobotics** [comment]
Merge Conflict for some reason? I downloaded the new suit.dmi file and added the new stuff before putting it back in but apparently that hasn't fixed it. Not sure why.

**Comment by AyyRobotics** [comment]
Also after looking into the issue more, again, and trying to determine why the map won't display or be loaded in the admin proc, I have a theory that wherever the game is deciding what map to load isn't in the awaymissionconfig.txt file, because in-game an admin got nothing when they tried to load an awaymission, when that wasn't the case in local testing, atleast jungleresort came up that time.

**Comment by AyyRobotics** [comment]
Good catch!

**Comment by AyyRobotics** [comment]
> One thing is that it seems that the cyborg update doesn't work for borgs, tried spawning it in, but no effects came

That's odd, last round someone got access to the techs and I got to be one of the cyborgs testing it, and it seemed to work fine? It may be that the Cargo module of cyborgs doesn't get anything via the upgrade since they weren't coded with that, which is something I'll probably fix in a PR after this one!

**Comment by AyyRobotics** [comment]
Tested the most recent version of the codebase and it was still having the issue of 'deprecated config settings', which makes me think my previous idea was right. Not sure what else to do, maybe if the PR is fully merged it'll update it, but I'm not sure.

**Comment by AyyRobotics** [comment]
Don't want to leave this abandoned or anything, but I'm not sure what else I can do if the map won't load due to external config issues. If someone knows how to fix it I'll get on it immediately, but otherwise I can just hope it'll fix itself on a full merge?

**Comment by MysteryMan21333** [comment]
@AyyRobotics Good morning, I was looking through your code I don't see too many issues besides someone in the discord saying they can't seem to locate the borg modules when researching the technology for the well, borg modules your PR present. As for the external errors in terms of it trying to delete shuttle code, its a common error for the codebase at the moment, don't worry about it.

**Comment by AyyRobotics** [comment]
> Ah, you're right, they were a cargo borg - also for the print problem, they probably didn't check in the robotics techfab so don't worry about it, last time I checked they were there! Biggest problem is the fact the gateway isn't loaded, which I fear might be due to a server-side issue, considering I made an update on the xenoarch .dmm file, it got pulled but server still uses the old one, for some reason.
> 
> We could test the "load away missions" verb, but even that one is broken sadly

I've looked into getting it tested via that, and it does work fine in local testing if you un-comment the filepath, though I was told nothing shows up in the admin panel, so you're right in that regard. My personal theory is that the server runs on a local copy of the config on the host computer, but I can't know that for sure. If that issue gets fixed though, everything else should work fine!

File: config/awaymissionconfig.txt [hunk 1]
```diff
@@ -17,4 +17,6 @@ _maps/RandomZLevels/away_mission/jungleresort.dmm
 #_maps/RandomZLevels/away_mission/caves.dmm
 #_maps/RandomZLevels/away_mission/snowdin.dmm
 #_maps/RandomZLevels/away_mission/research.dmm
-#_maps/RandomZLevels/away_mission/SnowCabin.dmm
\ No newline at end of file
+#_maps/RandomZLevels/away_mission/SnowCabin.dmm
+modular_splurt/_maps/away_mission/AGRComplexClock.dmm
+modular_splurt/_maps/away_mission/AGRComplexCult.dmm
```
Comment (BongaTheProto) [inline config/awaymissionconfig.txt:22]: Please comment these two to keep the repo configuration file without maps by default

**Review by BongaTheProto — CHANGES_REQUESTED** [review]

**Comment by AyyRobotics** [comment]
As per request I re-hashed the two complex files, but left jungleresort since that one was like that by default.

File: config/awaymissionconfig.txt (hunk 1 above)
Comment (AyyRobotics) [reply config/awaymissionconfig.txt]: Done! Left jungleresort though, since that one was like that prior.

**Comment by AyyRobotics** [comment]
Posting just to confirm that this isn't abandoned or anything, everything should be done and the requested changes were put in, so I'm just waiting at this point to see if the map loading issue will be solved if its merged, since I don't think there's anything I can do on my end.

**Comment by AyyRobotics** [comment]
> Too large to post

Hey, thanks for posting about it! I didn't know about these issues prior. I did check the discord a bit but I didn't see much criticism there, so I thought things were okay. I did test them locally but ran into similar issues that the map used to have regarding not being in rotation, so I had to force-load them which is probably why I didn't catch some of the issues with runes and such. You do make some good points here! I'll see what I can do about this, especially with the broken cult powers.

**Comment by AyyRobotics** [comment]
Started a first round of balance tweaks, I'll do some more testing tomorrow to look at specifics and see if anything more needs to be tweaked like the cult mobs and power generation!

**Comment by CSDSP** [comment]
There is far too much stuff in AGRComplex.dm that should be moved to other files. Especially stuff like the reagents and the stock parts, those effect stuff outside the away mission, and as such should be in files mirroring the regular path for reagents and stock parts.

They specifically effect stuff even without the away mission loaded, due to the possibility of getting empowered zinc with strange seeds, the primary way people have been getting these T6 parts on station.

**Comment by AyyRobotics** [comment]
Okay I might have forgotten how to fix merge conflicts, going to try and fix this

**Comment by AyyRobotics** [comment]
> Heyhey, hope all is good <3 There is one last thing I'd love to suggest, that is: It'd be really great if the Blood Cult map variation spawned with a single strand of Matrigrass, and the Clockwork one with a strand of sanguintare This is to make sure with keen exploration you can get both rewards (instead of creating T6 parts for the Blood Cult) Thank you very much <3

Also thought I'd reply to this!

I'm considering it, but there are some upsides and downsides to it? On one hand it'd mean explorers don't have to see the map and realize it's not the one with the rewards they may've wanted, but on the other removes a bit of the exclusivity that gives each map some of its identity as well as a little bit of balance? Going to have to think on this one!

**Comment by AyyRobotics** [comment]
> Heyhey, how did it go with this? Also, don't worry about balance, in the end it'd be just to make sure both maps give a good amount of rewards, and one isn't favoured over the other (just by adding 1 strand of matrigrass in the blood cult map, and 1 strand of sanguintare in the clockcult <3 )

You bring up a good point! Currently the rewards for the Blood Cult and Clock Cult are a little one-sided, what with the T6 Parts and all. I think I can make it work!

**Comment by AyyRobotics** [comment]
> Ah by the way been told one last thing: cultists should have exiles implant (sadly alot of people think they should antag and invade station

I probably will do that, although I worry it won't matter too much? If I remember right, implants are removable via surgery, so if there's more than one cultist they'd be able to get rid of them. Plus, I worry it'd make people frusturated, especially solo cultists who fight through the whole map only to reach the gate and realize they can't actually leave, or even multiple cultists who realize and have to double back in order to fix it. One thing I will do now though is update the spawn text so it's a bit clearer that they aren't supposed to be round antags!

**Review by BongaTheProto — CHANGES_REQUESTED** [review]
Please put the exile implants so we can merge this pr

**Comment by AyyRobotics** [comment]
> Please put the exile implants so we can merge this pr

Done Deal, gonna figure out how to do that and probably have it done within the hour

**Pull request merged**